llvm/lib/Support/Windows/Path.inc

```diff
@@ -785,9 +785,20 @@ std::error_code status(const Twine &path, file_status &result, bool Follow) {
 
   DWORD Flags = FILE_FLAG_BACKUP_SEMANTICS;
   if (!Follow) {
-    DWORD attr = ::GetFileAttributesW(path_utf16.begin());
-    if (attr == INVALID_FILE_ATTRIBUTES)
-      return getStatus(INVALID_HANDLE_VALUE, result);
+    DWORD attr;
+
+    // If getting file attributes fails due to a pending deletion, try
+    // again in a loop to avoid returning a misleading permission denied
+    // error.
+    for (int Retry = 200; Retry >= 0; --Retry) {
+      attr = ::GetFileAttributesW(path_utf16.begin());
+      if (attr != INVALID_FILE_ATTRIBUTES)
+        break;
+      std::error_code code = getStatus(INVALID_HANDLE_VALUE, result);
+      if (code != llvm::errc::delete_pending || !Retry)
+        return code;
+      ::Sleep(15);
```

Comment:
Am I reading this right, that this sleeps up to 3 seconds waiting for the condition to resolve?  That seems like sort of a long time to wait without the caller explicitly requesting it.

Reply:
Yes, you are reading this right. The loop here is expected to succeed within a few iterations, and if doesn't complete within the 3 second window, that's a good indication that there's probably something else awry.

But, of course, callers may be able to handle this failure case and prefer to avoid a 3 second penalty.

I'll test to see how many iterations are reached under a normal workload.

Reply:
Looking a bit more, I found that you can avoid hitting this error code in some cases by changing how you delete files: if you rename a file before deleting it, that frees the name immediately, instead of waiting for the remaining handles to be closed.  Does that apply here?

Reply:
To report the results of my test: in a lengthy compile with many (100k+) calls to the `status()` function, the loop only required extra iteration 3 times, with total number of iterations needed being 2, 3, and 3.

As for renaming the file before deleting it: the particular instance of this race I'm addressing is one where the file being `stat`'d is the destination file of another rename.

(for reference)
The call to `fs::rename`: https://github.com/apple/llvm-project/blob/32aa837fd7b392532b1180c63dace8ce10d0a5e6/clang/lib/Index/IndexUnitWriter.cpp#L401C3-L401C87

The call to `fs::status`: https://github.com/apple/llvm-project/blob/32aa837fd7b392532b1180c63dace8ce10d0a5e6/clang/lib/Index/IndexUnitWriter.cpp#L247

Several processes all race to rename temporary files to the same destination location, while one or more other processes race to `stat` this same destination location. So the rename-before-deleting change would be in `fs::rename` on the destination file, and it looks like this is already opportunistically applied: https://github.com/llvm/llvm-project/blob/0f17d9a28c40eebd42c83956e2a7b5186c1814d7/llvm/lib/Support/Windows/Path.inc#L513-L523

Reply:
The only way to get a delete_pending error is, as far as I can tell, to actually delete a file in-place.  If you rename a file, then delete it, the original name should get freed immediately after the rename.

My best theory for what's actually happening is that the "rename" is actually creating its own issue: somehow, the code for renaming a file on top of an existing file actually ends up deleting the existing file in-place, leading to delete_pending errors if some other process has the file open.  This seems like it should be possible to fix without sleeping...

Reply:
Yes, I suspect it's the `FILE_FLAG_DELETE_ON_CLOSE` flag passed here. Consider this test program which vaguely emulates two processes racing to `CreateFileW` at a common destination, and a third process trying to query this file:
```
#include <Windows.h>
#include <thread>
#include <cassert>

int main() {
    static LPCWSTR filePath = L"%TEMP%";
    std::thread createFileThread([]()
    {
        while (true)
        {
            HANDLE h = CreateFileW(filePath, GENERIC_READ | DELETE, 
                   FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, 
                   NULL, OPEN_EXISTING, 
                   FILE_ATTRIBUTE_NORMAL | FILE_FLAG_DELETE_ON_CLOSE, NULL);
            assert(h != NULL);
            Sleep(100);
            assert(CloseHandle(h));
            Sleep(100);
        }
    });

    std::thread createFileThread2([]()
    {
        while (true)
        {
            HANDLE h = CreateFileW(filePath, GENERIC_READ | DELETE, 
                   FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, 
                   NULL, OPEN_ALWAYS, 
                   FILE_ATTRIBUTE_NORMAL, NULL);
            assert(h != NULL);
            Sleep(100);
            assert(CloseHandle(h));
            Sleep(100);
        }
    });

    std::thread getFileAttributesThread([]()
    {
        while (true)
        {
            auto attributes = GetFileAttributesW(filePath);
            if (attributes == INVALID_FILE_ATTRIBUTES)
            {
                auto lastError = GetLastError();
                assert(lastError != ERROR_ACCESS_DENIED);
            }
        }
    });

    getFileAttributesThread.join();
}
```
After a few iterations of the loop, the assertion in the `getFileAttributesThread` fails.

There may be a way to change the rename logic to avoid this, but it still seems valuable to me for `fs::status` to be able to handle the case where a deletion is pending on the queried file.

Reply:
3 seconds is both a very long time to be waiting for a situation that isn't guaranteed to resolve (a file can be in a pending-delete state for an arbitrary length of time)... and also maybe not that long on a heavily loaded system, or a system with certain kinds of antivirus.  I'd much prefer to resolve the underlying condition in a deterministic way, and get rid of the loop.

Fixing the error code seems useful however we resolve this, though.

Reply:
I share your apprehension about sleeping in a loop like this. This wasn't the initial solution I was hoping for :/

Some alternate solutions considered that avoid the loop are:
1. in the case where we encounter `STATUS_DELETE_PENDING`, go ahead and return `no_such_file_or_directory`. The downside here is a stuck process could be holding on to the pending-deletion file handle, which would mean the return of `status` does not at all accurately reflect the state of the file. There is perhaps some precedent for this though, [Wine's implementation of `PathFileExistsW`](https://github.com/wine-mirror/wine/blob/1d3b312f225f79bec74d3d265128ead455467e2a/dlls/kernelbase/path.c#L2636) will return `false` if the queried path is pending deletion (indicated via `GetFileAttributesW(path) != INVALID_FILE_ATTRIBUTES`).
2. in the case where we encounter `STATUS_DELETE_PENDING`, return the new `errc::delete_pending` and require callers to handle this (potentially via looping). This avoids embedding a potentially expensive loop here, but for callers to handle this properly, they will have to do more or less the same thing (and I suspect this Windows-specific error condition just won't get handled most of the time; maybe that's okay).

Do these seem more palatable? There's some more discussion in the issue about alternate solutions: https://github.com/llvm/llvm-project/issues/89137

Reply:
Whether uses would want the loop probably depends on your assumptions. I mean, for most inputs/outputs of compilers, nothing else should be accessing the inputs/outputs in parallel.  So spitting out an error doesn't seem like a big deal, and waiting a few seconds is unlikely to actually resolve any issue.  You only run into this issue when you're intentionally racing with other processes... making such uses explicitly opt-in to doing something special seems fine.

Reply:
At the risk of [repeating myself](https://github.com/llvm/llvm-project/issues/89137#issuecomment-2077056218), I feel all this is caused by the lack of synchronization between processes that write the module units (in Apple's symbol index generation). The filesystem is used as a synchronization, but since you can't provide "scopes" like a mutex guard, you end up with this issue. Is that something that was considered in the application's design @z2oh ?

Reply:
I haven't worked on the symbol index generation code directly, so can't speak with conviction to the considerations taken in its design. My understanding is that explicit synchronization was anticipated to be more expensive than potentially duplicating work regenerating portions of the index.

At a high level, a compile job for a particular module will generate the symbol index for that module as well as all input (downstream) modules. There's a good chance that a downstream module is included in multiple upstream modules, so a different compile job may have already generated that symbol index file. There's a quick check (via `fs::status`) to see if that file has already been generated and is not stale, which avoids the duplicate work, but since this work is so inexpensive, it's not a big deal if some raciness messes up the underlying check and the index is regenerated unnecessarily.

My initial solution here was just to ignore the `permission_denied` errors and treat them as an indication that the index needed to generated, but I wanted to avoid leaking this Windows idiosyncrasy into platform agnostic code. Ideally this is accomplished by unifying the interface across platforms (i.e. the range of potential errors), but given the fundamental differences in filesystem implementation, it seems like there isn't a good way to do this without resorting to behavior like sleeping in a loop.

With all that said, I think returning just returning `errc::delete_pending` and handling this case downstream is reasonable. This avoids the suppression of legitimate permission errors, and is in theory not Windows-specific (though of course in practice it is).